### PR TITLE
ReClique: Default role to Virtual YMCA if no mapping (to master)

### PR DESCRIPTION
**Pull request by TomScarboroughYMCA**
**Related Issue/Ticket:** https://github.com/ymcatwincities/openy_gated_content/issues/122

**PLEASE CHECK BASE BRANCH FOR YOUR PR**
**ONLY urgent and approved fixes for point release should go to master branch**

https://github.com/ymcatwincities/openy_gated_content/issues/122

## Steps to test:

- Go to https://northpennymca-azure-stage.y.org (or equivalent Virtual Y site that uses ReClique CORE crm).
- Verify there are no Permission Mappings present in the Reclique authentication provider configuration.
- Create a new user in the system by logging in with an member's email address for that that is not already in the system
- example used for North Penn: nan_dec@yahoo.com. (Delete this user from Virtual Y site if already present)
- If this user is already present in People, delete that user from People to enable the test.
- Upon login, the user should have Virtual YMCA role assigned.

## Quality checks:

Please check these boxes to confirm this PR covers the following cases:

- Maintaining our upgrade path is essential. Check one or the other:
  - [ ] This PR  provides updates via `hook_update_N` or other means.
  - [X] No updates are necessary for this change.
- Front end fixes should be tested against all of the Open Y Themes.
  - [ ] Tested against Carnation
  - [ ] Tested against Lily
  - [ ] Tested against Rose
  - [X] This change does not contain front-end fixes.
- [ ] I have flagged this PR "Needs Review" or pinged the VY devs/QA
  team in Slack

### CODE REVIEW

**Comment by TomScarboroughYMCA** [comment]
retest

File: modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php [hunk 1]
```diff
@@ -66,14 +66,24 @@ public function onUserLogin(GCUserLoginEvent $event) {
             }
           }
           $user_membership = $event->extraData['member']['PackageName'];
```
Comment (retif) [inline modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php:68]: Seems like this variable is not used and can be removed
`$user_membership = $event->extraData['member']['PackageName'];`

Comment (froboy) [reply modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php:68]: @TomScarboroughYMCA ⬆️

Comment (TomScarboroughYMCA) [reply modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php]: Thanks - good catch

**Review by hamrant — DISMISSED** [review]
Code looks ok

**Comment by AnastasiiaPys** [comment]
retest

**Comment by AnastasiiaPys** [comment]
Testing failed. When I logged in with a test user I saw this error. Please check.
image
image

**Comment by hamrant** [comment]
cc @TomScarboroughYMCA

**Comment by TomScarboroughYMCA** [comment]
I'm not able to reproduce that error message in our staging environment.  Can you share with me the Reclique auth provider connection information you are using?

**Comment by AnastasiiaPys** [comment]
@TomScarboroughYMCA 
here are test url, un,pw and test email
image

**Comment by hamrant** [comment]
@TomScarboroughYMCA I think this not related to the OpenY version. Testing was on builds from this PR.

Based on the error, it's not hard to fix without reproducing:

https://github.com/fivejars/openy_gated_content/pull/283/files#diff-1f2c3ed7ae9e4c75c4d123ca20b5d5ac39b76dbe1fd7819da8e2283d01b51528R74 

Error on this line - `foreach ($user_memberships as $user_membership) {`, so we have case when `$user_memberships` can be `null` or `false` (not array) and we need to add `is_array` check for this variable before `foreach`. If not array - throw an error.

cc @AnastasiiaPys

**Comment by TomScarboroughYMCA** [comment]
Thanks for you feedback and suggestion, @hamrant. I added the is_array check, and re-tested with both reclique provider configurations.

**Comment by hamrant** [comment]
@TomScarboroughYMCA so, if there no memberships for the user, it will have a `virtual_y` role and access to VY, right?

I think if there no active memberships, no need to create an account and give access to VY.

@anpolimus @retif what do you think about this?

**Comment by TomScarboroughYMCA** [comment]
@hamrant - the scenario I'm coding for is when there are no permission mappings present. As a guide for this ReClique-based work, I referred to the Slack thread https://openy.slack.com/archives/C0131H70V38/p1607036956456100 and the work done in class GCAuthYUSAUserLoginSubscriber.php for the Y-USA provider.

@froboy - can you weigh in as well?

**Comment by hamrant** [comment]
@TomScarboroughYMCA I understand that :)

At least we need to improve this condition:

```php
          if (empty($active_roles)) {
            $active_roles = ['virtual_y'];
          }
```

To 

```php
          if (!empty($user_memberships) && empty($active_roles)) {
            $active_roles = ['virtual_y'];
          }
```

This will cover cases without permission mapping and not give access to customers without membership. But I think it better to return the error message `Your membership was expired` or something like this and skip the user creating.

**Comment by TomScarboroughYMCA** [comment]
Using `$event->extraData['Memberships'];` is incorrect, where the original code using `$event->extraData['member']['PackageName'];` is what is needed, along with the check for empty $active_roles following. This tested out the same as before, but without any errors written out to the Watchdog logs.

File: modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php [hunk 2]
```diff
@@ -66,14 +66,21 @@ public function onUserLogin(GCUserLoginEvent $event) {
...
```
Comment (hamrant) [inline modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php:68]: `$user_membership` can be a null or empty string? In this case, we will have `empty($active_roles)` and the same problem that was discussed before. I checked the code and we have the same situation in `openy_gc_auth_yusa`

cc @retif @anpolimus

**Review by hamrant — CHANGES_REQUESTED** [review]

File: modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php (hunk 2 above)
Comment (hamrant) [reply modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php:68]:
```php
          if ($user_membership && empty($active_roles)) {
            $active_roles = ['virtual_y'];
          }
```

Let's add this condition

Comment (TomScarboroughYMCA) [reply modules/openy_gc_auth/modules/openy_gc_auth_reclique/src/EventSubscriber/GCAuthReCliqueUserLoginSubscriber.php:68]: That works - user is getting authenticated with Virtual Y role assigned.

**Review by hamrant — APPROVED** [review]

**Comment by AnastasiiaPys** [comment]
Retested. Fixed)

**Comment by hamrant** [comment]
@TomScarboroughYMCA thanks for your patience

**Comment by TomScarboroughYMCA** [comment]
@hamrant - I thank you so much for your help and direction. I appreciate it.

**Pull request merged**